tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	// Secondarily select among other unused hosts.
+	for _, h := range t.hosts {
+		if _, ok := used[h]; ok {
+			continue
+		}
+		rv := tokenAwareBanditHostAndLatency{
+			host: h,
+		}
+		h.mu.RLock()
+		rv.hostInfo = h.hostInfo
+		rv.latency, rv.latencyOk = h.latency()
+		h.latency()
+		h.mu.RUnlock()
+		if !rv.hostInfo.IsUp() {
+			continue
+		}
+		if t.localDatacenter != "" {
+			rv.local = rv.hostInfo.DataCenter() == t.localDatacenter
+			if !rv.local {
+				rv.latency += t.remoteDatacenterPenalty
+			}
+		}
+		chooseFrom = append(chooseFrom, rv)
+	}
+	return chooseFrom
+}
+
+func (t *TokenAwareBanditHostPolicy) Pick(qry ExecutableQuery) NextHost {
+	iter := &tokenAwareBanditHostIterator{
+		policy: t,
+	}
+	t.mu.RLock()
+	iter.token, iter.replicas = t.findReplicas(qry)
+	t.mu.RUnlock()
+	iter.used = make(map[*tokenAwareBanditHost]struct{}, len(iter.replicas))
+	return iter.NextHost
+}
+
+type tokenAwareBanditHostIterator struct {
+	policy   *TokenAwareBanditHostPolicy
+	token    token
+	replicas []*HostInfo
+	used     map[*tokenAwareBanditHost]struct{}
+}
+
+func (hi *tokenAwareBanditHostIterator) selectHost(host tokenAwareBanditHostAndLatency) SelectedHost {
+	hi.used[host.host] = struct{}{}
+	return tokenAwareBanditSelectedHost{
+		host:  host.host,
+		info:  host.hostInfo,
+		token: hi.token,
+	}
+}
+
+// len(hosts) must be >1, otherwise this panics.
+func (hi *tokenAwareBanditHostIterator) randomHost(hosts []tokenAwareBanditHostAndLatency) SelectedHost {
+	index := rand.Intn(len(hosts) - 1)
+	return hi.selectHost(hosts[index])
+}
+
+func (hi *tokenAwareBanditHostIterator) NextHost() SelectedHost {
+	hosts := hi.policy.hostsToChooseFrom(hi.used, hi.replicas)
+
+	switch len(hosts) {
+	case 0:
+		return nil
+	case 1:
+		return hi.selectHost(hosts[0])
+	}
+
+	if rand.Float32() < hi.policy.explorationPortion {
+		// Exploration phase.
+		// We send some portion of queries to random nodes so that we have stats on the latency
+		// from all nodes.
+		return hi.randomHost(hosts)
+	}
+
+	// Otherwise we send the query to the best node (if it exists).
+	minIndex := -1
+	var minLatency time.Duration
+	for i := 0; i < len(hosts); i++ {
+		if !hosts[i].latencyOk {
+			// no data for this node.
+			continue
+		}
+		if minIndex == -1 || hosts[i].latency < minLatency {
+			minIndex = i
+			minLatency = hosts[i].latency
+		}
+	}
+
+	if minIndex == -1 {
+		// There is no host with min latency (we don't have data).
+		for i := range hosts {
+			if hosts[i].local {
+				return hi.selectHost(hosts[i])
+			}
+		}
+		return hi.randomHost(hosts)
+	}
+
+	return hi.selectHost(hosts[minIndex])
+}
+
+// tokenAwareBanditSelectedHost implements SelectedHost for TokenAwareBanditHostPolicy
+type tokenAwareBanditSelectedHost struct {
+	host      *tokenAwareBanditHost
+	info      *HostInfo
+	token     token
+	startTime time.Time
+}
+
+func (sh tokenAwareBanditSelectedHost) Info() *HostInfo {
+	return sh.info
+}
+
+func (sh tokenAwareBanditSelectedHost) Token() token {
```

Comment:
Why do we need to return non-exported type in the exported method?

Reply:
Unfortunately that's how gocql's `SelectedHost` interface looks like, we need to use the same signature:

https://github.com/kiwicom/gocql/blob/bd3bbd7282a06687131e84a77320c132005d9db1/policies.go#L314-L320

This is one of the reasons that v2 of the driver (or completely new driver from scratch) would be nice to have, we can't fix it in the current gocql version if we want to preserve backwards compatibility.

Reply:
I kind of lost in our forks and checked the  https://pkg.go.dev/github.com/gocql/gocql#SelectedHost
where haven't found this method, but it comes from github.com/scylladb/gocql